source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -185,5 +185,117 @@ LoadBalancerPtr ThreadAwareLoadBalancerBase::LoadBalancerFactoryImpl::create() {
+  //
+  // If weights are specified on the hosts, they are respected.
+  //
+  // This is an O(N) algorithm, unlike other load balancers. Using a lower `hash_balance_factor`
+  // results in more hosts being probed, so use a higher value if you require better performance.
+
+  if (normalized_host_weights_.empty()) {
+    return nullptr;
+  }
+
+  HostConstSharedPtr host = hashing_lb_ptr_->chooseHost(hash, attempt);
+  if (host == nullptr) {
+    return nullptr;
+  }
+  const double weight = normalized_host_weights_map_.at(host);
+  if (!isHostOverloaded(*host, weight)) {
+    ENVOY_LOG_MISC(debug,
+                   "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+                   "selected host #{} (attempt:1)",
+                   host->address()->asString());
+    return host;
+  }
+
+  // When a host is overloaded, we choose the next host in a random manner rather than picking the
+  // next one in the ring. The random sequence is seeded by the hash, so the same input gets the
+  // same sequence of hosts all the time.
+  const uint32_t num_hosts = normalized_host_weights_.size();
+  auto host_index = std::vector<uint32_t>(num_hosts);
+  for (uint32_t i = 0; i < num_hosts; i++) {
+    host_index[i] = i;
+  }
+
+  // Not using Random::RandomGenerator as it does not take a seed. Seeded RNG is a requirement
+  // here as we need the same shuffle sequence for the same hash every time.
+  // Further, not using std::default_random_engine and std::uniform_int_distribution as they
+  // are not consistent across Linux and Windows platforms.
+  const uint64_t seed = hash;
+  std::mt19937 random(seed);
+
+  // generates a random number in the range [0,k) uniformly.
+  auto uniform_int = [](std::mt19937& random, uint32_t k) -> uint32_t {
+    uint32_t x = k;
+    while (x >= k) {
+      x = random() / ((static_cast<uint64_t>(random.max()) + 1u) / k);
+    }
+    return x;
+  };
+
+  HostConstSharedPtr alt_host, max_weight_host = host;
+  double max_weight = hostWeight(max_weight_host);
+  for (uint32_t i = 0; i < num_hosts; i++) {
+    // The random shuffle algorithm
+    const uint32_t j = uniform_int(random, num_hosts - i);
+    std::swap(host_index[i], host_index[i + j]);
+
+    const uint32_t k = host_index[i];
+    alt_host = normalized_host_weights_[k].first;
+    if (alt_host == host) {
+      continue;
+    }
+
+    const double alt_host_weight = normalized_host_weights_[k].second;
+    if (!isHostOverloaded(*alt_host, alt_host_weight)) {
+      ENVOY_LOG_MISC(debug,
+                     "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+                     "selected host #{}:{} (attempt:{})",
+                     k, alt_host->address()->asString(), i + 2);
+      return alt_host;
+    }
+
+    const double alt_max_weight = hostWeight(alt_host);
+    if (alt_max_weight > max_weight) {
+      max_weight_host = alt_host;
+      max_weight = alt_max_weight;
+    }
```

Comment:
Naively, I would assume that you actually want to return the least overloaded host, not the one with the max weight. Why not just return the "amount overloaded" from `isHostOverloaded` and then use that for which host to pick if they are all overloaded? I might just not understand the algorithm.